receiver/httpcheckreceiver/scraper.go

```diff
@@ -23,50 +26,70 @@ var (
 )
 
 type httpcheckScraper struct {
-	client   *http.Client
+	clients  []*http.Client
 	cfg      *Config
 	settings component.TelemetrySettings
 	mb       *metadata.MetricsBuilder
 }
 
 // start starts the scraper by creating a new HTTP Client on the scraper
 func (h *httpcheckScraper) start(_ context.Context, host component.Host) (err error) {
-	h.client, err = h.cfg.ToClient(host, h.settings)
+	for _, target := range h.cfg.Targets {
+		client, clentErr := target.ToClient(host, h.settings)
+		if clentErr != nil {
+			err = multierr.Append(err, clentErr)
+		}
+		h.clients = append(h.clients, client)
+	}
 	return
 }
 
 // scrape connects to the endpoint and produces metrics based on the response
 func (h *httpcheckScraper) scrape(ctx context.Context) (pmetric.Metrics, error) {
-	if h.client == nil {
+	if h.clients == nil || len(h.clients) == 0 {
 		return pmetric.NewMetrics(), errClientNotInit
 	}
 
-	now := pcommon.NewTimestampFromTime(time.Now())
-
-	req, err := http.NewRequestWithContext(ctx, h.cfg.Method, h.cfg.Endpoint, http.NoBody)
-	if err != nil {
-		return pmetric.Metrics{}, err
+	var wg sync.WaitGroup
+	wg.Add(len(h.clients))
+	var mux sync.Mutex
+
+	for idx, client := range h.clients {
+		go func(targetClient *http.Client, targetIndex int) {
+			defer wg.Done()
+
+			now := pcommon.NewTimestampFromTime(time.Now())
+
+			req, err := http.NewRequestWithContext(ctx, h.cfg.Targets[targetIndex].Method, h.cfg.Targets[targetIndex].Endpoint, http.NoBody)
+			if err != nil {
+				h.settings.Logger.Error("failed to create request", zap.Error(err))
+				return
+			}
+
+			start := time.Now()
+			resp, err := targetClient.Do(req)
+			mux.Lock()
```

Comment:
I think I don't understand why we need this mutex. @srikanthccv would you be so kind and explain the potential problem and now introducing this mutex saves us from the problem?

Reply:
We scrape each endpoint in a separate go routine. There is a data race in generated_metrics `metricHttpcheckDuration.recordDataPoint`. This mutex is introduced for that reason. Let me know if there is an alternative way to solve this.

Reply:
I'm not sure if I see the race in the right place, but I think I see it :sweat_smile: The function [metricHttpcheckDuration.recordDataPoint](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/httpcheckreceiver/internal/metadata/generated_metrics.go#L29) calls [NumberDataPointSlice.AppendEmpty()](https://github.com/open-telemetry/opentelemetry-collector/blob/v0.81.0/pdata/pmetric/generated_numberdatapointslice.go#L81), which does this:

```go
	*es.orig = append(*es.orig, &otlpmetrics.NumberDataPoint{})
	return es.At(es.Len() - 1)
```

I think this is where the race can occur. We could have two goroutines execute the first line and then both returning the same (last) appended data point. Is this what you had in mind @srikanthccv ?

Reply:
@astencel-sumo Yes, it's the append.